### PR TITLE
Minimum Insertion to make string palindrome files added.

### LINKED ISSUE #1576: I would like to work on Minimum insertion in string to make palindrome using DP.

### PR DESCRIPTION
**Pull request by Shiv1202**
Fixes #1576

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [ ] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Write a program for Minimum Insertion In a string to make it Palindrome.
- Naive method and Optimized Both methods are added.

#### Languages Used:
- Python 3
- 

#### Files Added:
- minimum_insertion_to_make_palindrome_basic.py
- minimum_insertion_to_make_palindrome_DP.py 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by Shiv1202** [comment]
@singh-shreya6 Ma'am please review.

**Comment by SakshiShreya** [comment]
Please complete the checklist properly.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        return 0
+    if l == h - 1:
+        return 0 if str[l] == str[h] else 1
+    # Check if the first and last char are same 
+    # on the basis of the comparison result, deside
+    # which subproblem(s) to call.
+    if str[l] == str[h]:
+        return find_min_insertion(str,l+1,h-1)
+    else:
+        return min(find_min_insertion(str,l,h-1), find_min_insertion(str,l+1,h)) + 1
```
Comment (SakshiShreya) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py]: Please split this in two lines, one line can't contain more than 80 characters

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# Main Function.
+def main():
+    s = "abcd"
+    n = len(s) - 1
+    print(find_min_insertion(s, 0, n))
+
+# Driver Code.
+if __name__ == "__main__":
+    main()
```
Comment (SakshiShreya) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py]: Add a new line at the end of file

Comment (SakshiShreya) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py]: Mention a sample input and output at the end of file.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+Time Complexity : O(n^2)
+Space Complexity : O(n^2)
+--------------------------------------- """
+# Function for Findind Min.
+def Min(a, b):
+    """
+        This Function is return minimum value B/w a and b.
+    """
+    return min(a, b)
+
```
Comment (SakshiShreya) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py]: Remove the extra lines between functions. Only one new-line should be there between two functions.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    
+# Main Function.
+def main():
+    s = "abcd"
+    x = find_min_insertion_DP(s, len(s))
+    print(x)
+
+# Driver Code.
+if __name__ == "__main__":
+    main()
```
Comment (SakshiShreya) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py]: Add a new line at the end of file and provide sample input and output

**Comment by Shiv1202** [comment]
Ma'am, I made the requested changes.
Please review it.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Python Program to find min Insertion in a string to make it Palindrome.
```
Comment (jainaman224) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py]: Incorrect file name.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+# Driver Code.
+if __name__ == "__main__":
+    main()
+
+"""
+    [Simple Input]
+      s = 'dcbabcd'
+    [Simple Output]
+      min_insertion = 3
+"""
```
Comment (jainaman224) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py]: Add new line at the end of file.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+"""
+    [Simple Input 1]
+      s = 'car'
+    [Simple Output 1]
+      min_insertion = 2
+    [Simple Input 2]
+      s = 'radar'
+    [Simple Output 2]
+      min_insertion = 0
+"""
```
Comment (jainaman224) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py]: Add new line at the end of file.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    # which subproblem(s) to call.
+    if str[l] == str[h]:
+        return find_min_insertion(str,l+1,h-1)
+    else:
+        val=min(find_min_insertion(str,l,h-1),find_min_insertion(str,l+1,h))
+        return val+1
+
+# Main Function.
+def main():
+    s = "abcd"
```
Comment (jainaman224) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_basic.py]: Take input from user.

File: Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            if s[l] == s[h]:
+                t[l][h] = t[l+1][h-1]
+            else:
+                t[l][h] = (Min(t[l][h-1], t[l+1][h]) + 1)
+            l += 1
+    return t[0][n - 1]
+ 
+# Main Function.
+def main():
+    s = "car"  # Input String
```
Comment (jainaman224) [inline Minimum_insertion_to_make_Palindrome/minimum_insertion_to_make_palindrome_optimize.py]: Take input from user.

**Comment by somya-kapoor** [comment]
@Shiv1202 Are you working on it?

**Comment by somya-kapoor** [comment]
@Shiv1202 Are you working on it?

**Comment by raghav-dalmia** [comment]
You are unresponsive for more than 11days. So, I am closing this PR.

**Pull request closed without merging**